### PR TITLE
[NPUW] Native bidirectional-encoder text-embedding path on NPU 

### PR DESCRIPTION
**Pull request by dylanneve1**
### Details:
 - Add support for encoder embedding models in NPUW pipeline, tested with bge-small-zh-v1.5, xiaobu-embedding-v2 and Conan-embedding-v1, all BERT based embedding models

### Tickets:
 - EISW-192888

### AI Assistance:
 - *AI assistance used: no / yes*
 - *If yes, summarize how AI was used and what human validation was performed (build/tests/manual checks).*

### CODE REVIEW

**Review by dmatveev — APPROVED** [review]
To me this change looks legit - mainly it introduces a new path that shouldn't break the existing behavior, and the new functionality is neatly isolated from the rest of the code.

Well done @dylanneve1 !

@AlexanderKalistratov  @AsyaPronina I'd still expect you to look at this PR before we merge it.

File: src/plugins/intel_npu/src/plugin/npuw/llm_compiled_model.cpp [hunk 1]
```diff
@@ -1355,9 +1399,11 @@ void ov::npuw::LLMCompiledModel::serialize(std::ostream& raw_stream, const ov::n
         write_model_meta(raw_stream);
     }
 
-    // Serialize bank name
-    const auto& kv_bank = m_kvcache_compiled->get_weights_bank();
+    // Serialize bank name.
+    // Encoder embedding models are prefill-only (no kvcache/generate compiled model), so use the
+    // prefill model's weight bank; otherwise prefill and kvcache must share the same bank.
     const auto& p_bank = m_prefill_compiled->get_weights_bank();
+    const auto& kv_bank = m_kvcache_compiled ? m_kvcache_compiled->get_weights_bank() : p_bank;
```
Comment (dmatveev) [inline src/plugins/intel_npu/src/plugin/npuw/llm_compiled_model.cpp]: What difference does it make here?

Comment (dylanneve1) [reply src/plugins/intel_npu/src/plugin/npuw/llm_compiled_model.cpp]: Encoder embedding models are prefill only, so `m_kvcache_compiled` is null for them, without this change it would crash

File: src/plugins/intel_npu/src/plugin/npuw/embedding/encoder_embedding_infer_request.hpp [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+#include "../llm_compiled_model.hpp"
+#include "../llm_infer_base_request.hpp"
+#include "openvino/core/descriptor/output.hpp"
+
+namespace ov {
+namespace npuw {
+
+// Infer request for non-autoregressive bidirectional encoder (e.g. BERT) text-embedding models.
+// A single static forward over the whole (right-padded) sequence; no KV cache, no chunking, no
+// position_ids (absolute positions are internal to the model). Compare EmbeddingInferRequest,
```
Comment (AsyaPronina) [inline src/plugins/intel_npu/src/plugin/npuw/embedding/encoder_embedding_infer_request.hpp]: Just a question for future: Is it common for all BERT-like models to have absolute positions?

File: src/plugins/intel_npu/src/plugin/npuw/embedding/encoder_embedding_infer_request.cpp [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+
+    m_prefill_request->infer();
+
+    // Copy the valid (front) rows of the hidden state into the right-sized output tensor; the
+    // remaining padded rows stay zeroed. Downstream pooling uses attention_mask to ignore them.
+    uu::fill_tensor_bytes(m_prefill_output, 0u);
+    auto output_tensor = m_prefill_request->get_tensor(m_prefill_request->get_outputs()[0]);
+    auto src = uu::make_tensor_slice(output_tensor, layer_ids::INPUT_IDS_SEQ_LEN_DIM, 0, prompt_len);
+    auto dst = uu::make_tensor_slice(m_prefill_output, layer_ids::INPUT_IDS_SEQ_LEN_DIM, 0, prompt_len);
+    uu::copy_tensor_by_dim(src, dst, layer_ids::INPUT_IDS_SEQ_LEN_DIM, layer_ids::INPUT_IDS_SEQ_LEN_DIM);
```
Comment (AsyaPronina) [inline src/plugins/intel_npu/src/plugin/npuw/embedding/encoder_embedding_infer_request.cpp]: Why do we need a copy to custom tensor here, can't we just reuse the ` m_prefill_request->get_tensor(m_prefill_request->get_outputs()[0])` and assign result to the `m_prefill_output`? Do we need to allocate `m_prefill_output` if we have ` m_prefill_request->get_tensor(m_prefill_request->get_outputs()[0])`?

File: src/plugins/intel_npu/src/plugin/npuw/embedding/prepare_encoder_embedding_model.cpp [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+    // as-is (ReshapeToStatic later fixes the shapes to [1, L]); here we only sanity-check that
+    // there is no autoregressive KV cache and re-validate shapes.
+    for (const auto& param : model->get_parameters()) {
+        for (const auto& name : param->get_output_tensor(0).get_names()) {
+            OPENVINO_ASSERT(name.find("past_key_values") == std::string::npos,
+                            "PrepareEncoderEmbeddingModel: unexpected autoregressive KV-cache input '",
+                            name,
+                            "' in an encoder embedding model.");
+        }
+    }
```
Comment (AsyaPronina) [inline src/plugins/intel_npu/src/plugin/npuw/embedding/prepare_encoder_embedding_model.cpp]: Do we really need an additional transformation that does only assertion checks? Can this be done in just some function? Why the pass variant was chosen? I believed that passes usually does some transformations to the model or find subgraph nodes for future transformations...

File: src/plugins/intel_npu/src/plugin/npuw/llm_compiled_model.cpp [hunk 5]
```diff
@@ -816,8 +821,38 @@ ov::npuw::LLMCompiledModel::LLMCompiledModel(const std::shared_ptr<ov::Model>& m
...
-        ov::npuw::util::PrepareTextEmbeddingModel(seq_len_dim).run_on_model(kvcache_model);
+        // Distinguish autoregressive (Qwen3-Embedding-style) from non-autoregressive bidirectional
+        // encoders (BERT: xiaobu, bge, Conan). The latter cannot be reconstructed into a prefill/KV
+        // model and must run as a single static forward. The check keys on the KV-cache concat
+        // pattern on the SDPA key input; log the verdict at INFO so a misclassified model (e.g.
+        // an unusual decoder topology routed to the encoder path) is diagnosable from the logs.
+        m_is_encoder_embedding = ov::npuw::util::is_encoder_embedding_model(kvcache_model);
+        LOG_INFO("Text-embedding model classified as "
+                 << (m_is_encoder_embedding ? "bidirectional encoder: single-forward, prefill-only path"
+                                            : "autoregressive decoder: reconstructed prefill/KV path"));
```
Comment (AsyaPronina) [inline src/plugins/intel_npu/src/plugin/npuw/llm_compiled_model.cpp]: Does autoregressive decode has generate path or what does KV path mean here? I believe it also has only prefill path, but has causal mask, so can utilize chunked prefill mode, where KV-Cache is required?

File: src/plugins/intel_npu/src/plugin/npuw/llm_compiled_model.cpp [hunk 6]
```diff
@@ -816,8 +821,38 @@ ov::npuw::LLMCompiledModel::LLMCompiledModel(const std::shared_ptr<ov::Model>& m
...
+            LOG_DEBUG("Encoder (bidirectional) text-embedding model rebuild");
+            // A bidirectional encoder attends over the whole sequence at once; chunked prefill is
+            // semantically invalid for it. Force a single whole-sequence forward.
+            m_use_chunk_prefill = false;
+            // The encoder uses a learned absolute position table of size max_position_embeddings.
+            // The static sequence length must not exceed it, otherwise the position embedding
+            // (clamped to the table) won't broadcast against the token embedding and compilation
+            // fails. Clamp here so the default NPUW_LLM_MAX_PROMPT_LEN (which is sized for LLMs and
+            // is typically larger, e.g. 1024) does not break a 512-position BERT out of the box.
+            if (auto max_pos = ov::npuw::util::get_max_position_embeddings(kvcache_model)) {
```
Comment (AsyaPronina) [inline src/plugins/intel_npu/src/plugin/npuw/llm_compiled_model.cpp:836]: Do we need to adjust `NPUW_LLM_MAX_PROMPT_LEN` here, don't we need to do it on GenAI side and leave only assert here?

File: src/plugins/intel_npu/src/plugin/npuw/llm_compiled_model_utils.cpp [hunk 7]
```diff
@@ -4,10 +4,77 @@
...
 }
+
+bool ov::npuw::util::is_encoder_embedding_model(const std::shared_ptr<ov::Model>& model) {
+    // Mirror of the discriminator used by ReConstructEmbeddingModel::check_kv_concat_nodes
+    // (embedding/prepare_embedding_model.cpp): the autoregressive embedding path requires the
+    // key input of SDPA to come through a Reshape <- Broadcast (GQA KV-cache concat) pattern.
+    // A bidirectional encoder (BERT) has SDPA but lacks this pattern, so it is NOT autoregressive.
+    auto has_kv_concat_pattern = [](const std::shared_ptr<ov::Node>& sdpa) -> bool {
+        // Key input is at index 1: Concat -> Broadcast -> Reshape -> SDPA.
+        // Must match ReConstructEmbeddingModel::check_kv_concat_nodes exactly, including the
```
Comment (AsyaPronina) [inline src/plugins/intel_npu/src/plugin/npuw/llm_compiled_model_utils.cpp]: Why does this comment exist: are there cases where `Broadcast -> Reshape -> SDPA` exists without `Concat`?

File: src/plugins/intel_npu/src/plugin/npuw/llm_compiled_model_utils.cpp [hunk 8]
```diff
@@ -4,10 +4,77 @@
...
+    // Convert/decompression nodes in between.
+    for (const auto& op : model->get_ops()) {
+        if (!ov::is_type<ov::op::v8::Gather>(op)) {
+            continue;
+        }
+        if (op->get_friendly_name().find("position_embeddings") == std::string::npos) {
+            continue;
+        }
+        auto src = op->input_value(0);  // the position-embedding weight (possibly behind Convert)
+        for (int hops = 0; hops < 4 && src.get_node(); ++hops) {
```
Comment (AsyaPronina) [inline src/plugins/intel_npu/src/plugin/npuw/llm_compiled_model_utils.cpp]: This seems a bit fragile... Don't we want to use a real pattern-matching here to find `Gather` node?

Comment (AsyaPronina) [reply src/plugins/intel_npu/src/plugin/npuw/llm_compiled_model_utils.cpp]: Or at list use the real intermediate nodes in the graph while traversing, explicitly stating them (`Convert`) and so on?

File: src/plugins/intel_npu/src/plugin/npuw/embedding/encoder_embedding_infer_request.cpp [hunk 9]
```diff
@@ -0,0 +1,118 @@
...
+#include "../logging.hpp"
+#include "../util.hpp"
+
+ov::SoPtr<ov::ITensor> ov::npuw::EncoderEmbeddingInferRequest::create_prefill_output_tensor() {
+    const auto& out_port = m_prefill_request->get_outputs()[0];
+    auto out_shape = out_port.get_shape();
+    auto& kvcache_desc = m_npuw_llm_compiled_model->m_kvcache_desc;
+
+    auto prefill_shape = ov::Shape(out_shape);
+    prefill_shape[layer_ids::INPUT_IDS_SEQ_LEN_DIM] = kvcache_desc.max_prompt_size;
```
Comment (AlexanderKalistratov) [inline src/plugins/intel_npu/src/plugin/npuw/embedding/encoder_embedding_infer_request.cpp]: I'm not sure why we need to do this? Isn't our infer request already statically shaped with output shape with correct value?

And we also allocating a new tensor. Can't we reuse m_prefill_request output buffer?

**Review by AlexanderKalistratov — COMMENTED** [review]
I basically agree with @AsyaPronina comments

File: src/plugins/intel_npu/src/plugin/npuw/llm_compiled_model.cpp (hunk 6 above)
Comment (AlexanderKalistratov) [reply src/plugins/intel_npu/src/plugin/npuw/llm_compiled_model.cpp:836]: I think it is fine. BERT has fixed size of 256 or 512.
We probably should ignore `NPUW_LLM_MAX_PROMPT_LEN` at all. And infer it from model itself.

**Review by AlexanderKalistratov — APPROVED** [review]

File: src/plugins/intel_npu/src/plugin/npuw/embedding/encoder_embedding_infer_request.cpp (hunk 3 above)
Comment (dylanneve1) [reply src/plugins/intel_npu/src/plugin/npuw/embedding/encoder_embedding_infer_request.cpp]: Thanks, done, the copy is gone

**Comment by dylanneve1** [comment]
https://cje-ir-prod01.devtools.intel.com/sai-npu-experience/job/Staging/job/dneve/job/Validate/136/Validation_20report/

File: src/plugins/intel_npu/src/plugin/npuw/embedding/encoder_embedding_infer_request.hpp (hunk 2 above)
Comment (dylanneve1) [reply src/plugins/intel_npu/src/plugin/npuw/embedding/encoder_embedding_infer_request.hpp]: Most of the classic BERT family does (bert, roberta, minilm, bge and friends all use a learned absolute position table). Some newer encoders switched to RoPE or ALiBi instead. The path itself doesn't depend on it though, the table is only probed to clamp the static sequence length, and if there's no table the clamp is skipped.

File: src/plugins/intel_npu/src/plugin/npuw/embedding/prepare_encoder_embedding_model.cpp (hunk 4 above)
Comment (dylanneve1) [reply src/plugins/intel_npu/src/plugin/npuw/embedding/prepare_encoder_embedding_model.cpp]: Agreed, it didn't need to be a pass. It's a plain validate_encoder_embedding_model() function now and the pass is gone.

File: src/plugins/intel_npu/src/plugin/npuw/llm_compiled_model.cpp (hunk 5 above)
Comment (dylanneve1) [reply src/plugins/intel_npu/src/plugin/npuw/llm_compiled_model.cpp]: You're right, both embedding flavours only ever prefill. By KV path I meant the reconstructed KV-cache inputs plus the causal mask, which is what lets the autoregressive one prefill in chunks. The encoder has neither, so it runs one whole-sequence forward. The log line and the comment are reworded to say exactly that now.

File: src/plugins/intel_npu/src/plugin/npuw/llm_compiled_model_utils.cpp (hunk 8 above)
Comment (dylanneve1) [reply src/plugins/intel_npu/src/plugin/npuw/llm_compiled_model_utils.cpp]: Reworked into a pattern matcher

File: src/plugins/intel_npu/src/plugin/npuw/embedding/encoder_embedding_infer_request.cpp (hunk 9 above)
Comment (dylanneve1) [reply src/plugins/intel_npu/src/plugin/npuw/embedding/encoder_embedding_infer_request.cpp]: Reworked, the extra tensor and the copy are gone. The request now returns the compiled model's own output tensor directly and just zeroes the padding rows in place after infer.

File: src/plugins/intel_npu/src/plugin/npuw/llm_compiled_model.cpp [hunk 10]
```diff
@@ -812,8 +812,45 @@ ov::npuw::LLMCompiledModel::LLMCompiledModel(const std::shared_ptr<ov::Model>& m
...
+        // autoregressive embedder (Qwen3-Embedding-style) has a causal mask and a KV cache to
+        // reconstruct, which is what lets it prefill in chunks. A bidirectional encoder (BERT:
+        // xiaobu, bge, Conan) attends over the whole sequence at once and has no KV cache to
+        // rebuild, so it runs as a single static forward instead. The check keys on the KV-cache
+        // concat on the SDPA key input; log the verdict at INFO so a model that lands on the wrong
+        // side of it (an unusual decoder topology, say) is diagnosable from the logs.
+        m_is_encoder_embedding = ov::npuw::util::is_encoder_embedding_model(kvcache_model);
+        LOG_INFO("Text-embedding model classified as "
+                 << (m_is_encoder_embedding ? "bidirectional encoder: whole-sequence forward, no KV cache"
+                                            : "autoregressive decoder: causal mask and reconstructed KV cache"));
```
Comment (AsyaPronina) [inline src/plugins/intel_npu/src/plugin/npuw/llm_compiled_model.cpp:825]: `autoregressive encoder`

File: src/plugins/intel_npu/src/plugin/npuw/llm_compiled_model_utils.cpp (hunk 8 above)
Comment (AsyaPronina) [reply src/plugins/intel_npu/src/plugin/npuw/llm_compiled_model_utils.cpp]: Very cool, thanks a lot!

**Review by AsyaPronina — APPROVED** [review]
